File: packages/python-sdk/src/treedb_sdk/pagination.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Generic, TypeAlias, TypeVar


T = TypeVar("T")
TreeDbCursor: TypeAlias = str


@dataclass(frozen=True)
class TreeDbPage(Generic[T]):
    items: list[T]
    next_cursor: str | None = None
    has_more: bool | None = None
    cursor: str | None = None
    limit: int | None = None


def create_page(
    items: list[T],
    *,
    next_cursor: str | None = None,
    has_more: bool | None = None,
    cursor: str | None = None,
    limit: int | None = None,
) -> TreeDbPage[T]:
    return TreeDbPage(items=items, next_cursor=next_cursor, has_more=has_more, cursor=cursor, limit=limit)


def is_treedb_page(value: Any) -> bool:
    return isinstance(value, TreeDbPage) or (isinstance(value, dict) and isinstance(value.get("items"), list))


def get_next_cursor(page: TreeDbPage[Any]) -> str | None:
    return page.next_cursor

File: packages/ts-sdk/src/treedb/client/pagination.ts
import type { TreeDbPage } from '../types/index.js';

export function createPage<T>(input: TreeDbPage<T>): TreeDbPage<T> {
  return {
    items: input.items,
    nextCursor: input.nextCursor,
    hasMore: input.hasMore,
    cursor: input.cursor,
    limit: input.limit
  };
}

export function isTreeDbPage(value: unknown): value is TreeDbPage<unknown> {
  return typeof value === 'object' && value !== null && Array.isArray((value as TreeDbPage<unknown>).items);
}

export function getNextCursor<T>(page: TreeDbPage<T>): string | undefined {
  return page.nextCursor;
}
